fecgov/fec-cms
# Raising / spending back button fix

Added functionality to reflect the user's current choices in the URL, and restore those when using the browser history to return to the page.

## Summary
The pull-downs' states weren't being pushed to the URL so changing the cycle or office, clicking a candidate, then going back to the previous page (from the candidate details) reloaded the default version of the raising or spending data and didn't reflect the user's most recent settings.

Changing either the cycle or office pull-downs now pushes those updated values to the URL.

- Resolves [#2066](https://github.com/fecgov/fec-cms/issues/2066)

## Impacted areas of the application
`top-entities.js`:
- I added `uri.js` to handle url updates.
- added `office` to `baseQuery` in `init`
- borrowed `pushStateToURL()` from elsewhere; would like to streamline
- added `this.pushStateToURL()` to both `handleCycleChange()` and `handleCategoryChange()`

`views.py`:
- near-identical updates to both `raising()` and `spending()`
- now overriding `top_category` if the URL contains a (valid) value for `list`
- since cycle can now come from the URL, forced it/them to `int`

## Screenshots
Check out the URL changes:
![default](https://user-images.githubusercontent.com/26720877/52073643-00106380-2556-11e9-9437-20f0399b245d.png)
![with new url vars](https://user-images.githubusercontent.com/26720877/52073654-04d51780-2556-11e9-8481-0851fe3cc022.png)


## How to test
- Go to http://localhost:8000/data/raising-bythenumbers
- Change the cycle (remember what it was and how you changed it)
- Change the office pull-down (remember what it was and how you changed it)
- Look at the list of candidates (remember general names, order, and values)
- Click a candidate's name
- Go back (browser back button, keyboard cmd ⬅️, Android ⬅️, etc)
- Verify the cycle, office, and candidates are exactly how they were when you clicked the candidate's name
- ~~*Bonus*: change both the cycle and office values multiple times, click a candidate, and then go back several times to make sure each step through history is correct~~ Coming in a following ticket
- Do the same for http://localhost:8000/data/spending-bythenumbers/?cycle=2012&list=P

TESTING CONCERNS:
I'd like to make sure this change doesn't break something anywhere else these components are used. Where all might that be?
____
